processor/src/advice/mod.rs

```diff
@@ -7,188 +7,113 @@ use vm_core::{
-    pub fn read_tape_double(&mut self) -> Result<[Word; 2], ExecutionError> {
-        let word0 = self.read_tapew()?;
-        let word1 = self.read_tapew()?;
-
-        Ok([word0, word1])
-    }
+    fn read_tape_dw(&mut self) -> Result<[Word; 2], ExecutionError>;
 
     /// Writes the provided value at the head of the advice tape.
-    pub fn write_tape(&mut self, value: Felt) {
-        self.tape.push(value);
-    }
+    fn write_tape(&mut self, value: Felt);
 
-    /// Retrieves a list of elements from a key-value map for the specified key, reverses it, and
-    /// writes the reversed list at the head of the advice tape. This way, the first element in the
-    /// list is located at the head of the advice tape.
+    /// Fetch a keyed tape from the values map, reversing and appending it to the advice tape.
+    ///
+    /// Note: this operation shouldn't consume the map element so it can be called multiple times
+    /// for the same key.
+    ///
+    /// # Example
+    /// Given an advice map `[a,b,c]`, and a map `x |-> [d,e,f]`, a call `write_tape_from_map(x)`
+    /// will result in `[a,b,c,f,e,d]` for the advice tape, and will preserve `x |-> [d,e,f]`.
     ///
     /// # Errors
     /// Returns an error if the key was not found in a key-value map.
-    pub fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError> {
-        let values = self
-            .values
-            .get(&key.into_bytes())
-            .ok_or(ExecutionError::AdviceKeyNotFound(key))?;
-        for &elem in values.iter().rev() {
-            self.tape.push(elem);
-        }
-
-        Ok(())
-    }
-
-    /// Inserts a list of elements to the advice map with the top four elements of the stack as
-    /// the key.
+    fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError>;
+
+    /// Maps a key to a values list to be yielded by `write_tape_from_map`.
     ///
     /// # Errors
     /// Returns an error if the key is already present in the advice map.
-    pub fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), ExecutionError> {
-        match self.values.insert(key.into_bytes(), values) {
-            None => Ok(()),
-            Some(_) => Err(ExecutionError::DuplicateAdviceKey(key)),
-        }
-    }
+    fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), ExecutionError>;
 
     // ADVISE SETS
     // --------------------------------------------------------------------------------------------
 
-    /// Returns true if the advice set with the specified root is present in this advice provider.
-    #[cfg(test)]
-    pub fn has_advice_set(&self, root: Word) -> bool {
-        self.sets.contains_key(&root.into_bytes())
-    }
-
-    /// Returns a node at the specified index in a Merkle tree with the specified root.
+    /// Returns a node/leaf for the given depth and index on a Merkle tree with the given root.
```

Comment:
nit: should it be "in a Merkle tree" rather than "on a Merkle tree"?

Reply:
True!